srk/src/polynomialCone.ml

```diff
@@ -0,0 +1,310 @@
+
+type pvutil = {
+  mono_map: int MonomialMap.t;
+  ordered_mono_list: Monomial.t list;
+}
+
+let monomial_map_of_polynomial p =
+  let mono_map = MonomialMap.empty in
+  BatEnum.fold (fun m (coeff, monomial) ->
+      MonomialMap.add monomial coeff m)
+    mono_map
+    (QQXs.enum p)
+
+let pvutil_of_polys polys =
+  let map =
+    BatList.fold (fun map poly ->
+        let f _ a b =
+          match a, b with
+          | Some a, _ -> Some a
+          | _, Some b -> Some b
+          | None, None -> assert false
+        in
+        MonomialMap.merge f map (monomial_map_of_polynomial poly))
+      MonomialMap.empty
+      polys
+  in
+  let ordered_mono_map =
+    BatEnum.foldi
+      (fun i (monomial, _) map -> MonomialMap.add monomial i map)
+      MonomialMap.empty
+      (MonomialMap.enum map)
+  in
+  let ordered_mono_list =
+    BatList.of_enum (MonomialMap.enum ordered_mono_map) |> BatList.map (fun (m, _) -> m) in
+  {
+    mono_map = ordered_mono_map;
+    ordered_mono_list = ordered_mono_list;
+  }
+
+let pvutil_merge pvutil poly =
+  let mono_map = BatEnum.fold (fun m (_, monomial) ->
+      MonomialMap.add monomial 1 m)
+      pvutil.mono_map
+      (QQXs.enum poly)
+  in
+  let ordered_mono_map =
+    BatEnum.foldi
+      (fun i (monomial, _) map -> MonomialMap.add monomial i map)
+      MonomialMap.empty
+      (MonomialMap.enum mono_map)
+  in
+  let ordered_mono_list =
+    BatList.of_enum (MonomialMap.enum ordered_mono_map) |> BatList.map (fun (m, _) -> m) in
+  {
+    mono_map = ordered_mono_map;
+    ordered_mono_list = ordered_mono_list;
+  }
+
+let monomial_map_of_cone c =
+  BatList.fold (fun map poly ->
+      let f _ a b =
+        match a, b with
+        | Some a, _ -> Some a
+        | _, Some b -> Some b
+        | None, None -> assert false
+      in
+      MonomialMap.merge f map (monomial_map_of_polynomial poly))
+    MonomialMap.empty
+    c
+
+let vec_of_poly p pvutil =
+  let mono_map = pvutil.mono_map in
+  let v = V.zero in
+  BatEnum.fold (fun v (coeff, monomial) ->
+      let idx = MonomialMap.find monomial mono_map in
+      V.add_term coeff idx v)
+    v
+    (QQXs.enum p)
+
+let poly_of_vec vec pvutil =
+  let ordered_mono_list = pvutil.ordered_mono_list in
+  BatEnum.fold (fun p (coeff, index) ->
+      if index > -1 then
+        QQXs.add_term coeff (BatList.nth ordered_mono_list index) p
+      else QQXs.add_term coeff Monomial.one p)
+    (QQXs.zero)
+    (Linear.QQVector.enum vec)
+
+let polyhedron_of_cone cone_generators pvutil =
+  let vec_generators = BatList.map
+      (fun p -> vec_of_poly p pvutil)
+      cone_generators
+  in
+  let generators_enum = BatEnum.map
+      (fun v -> (`Ray, v))
+      (BatList.enum vec_generators)
+  in
+  BatEnum.push generators_enum (`Vertex, V.zero);
+  let dim = MonomialMap.cardinal pvutil.mono_map in
+  Polyhedron.of_generators dim generators_enum
+
+let find_implied_zero_polynomials polys basis =
+  logf "finding implied zero polynomials";
```

Comment:
If you add too much logging information, the useful stuff gets hidden.  Try to only commit informative log info.